### ISSUE TITLE
Legend positioning behavior change from matplotlib  {2.1.0 -> 2.2.2}

### ISSUE THREAD

**Issue by deepak18** [matplotlib/matplotlib#11126]
### Bug report

**Bug summary**
The behavior of positioning a legend outside the plot by using the figure transformation has changed, by shortening the size of plot and leaving too much space between plot and legend.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

# Fixing random state for reproducibility
np.random.seed(19680801)

N = 50
x = 15*np.random.rand(N)
y = 15*np.random.rand(N)
colors = np.random.rand(N)

fig, ax = plt.subplots(figsize=(8,6))

ax.scatter(x, y, c=colors, alpha=0.5, label='Scatter plot')
ax.plot([0, 15],[0, 15], color='r', linestyle='-', linewidth=2.5, label = 'Straight line')

ax.legend(bbox_to_anchor=(1, 0), loc="upper right",
           bbox_transform=fig.transFigure, columnspacing=0.5, ncol=2)

fig.tight_layout()
```

**Actual outcome**
![output_2 2 2](https://user-images.githubusercontent.com/5120865/39248486-33e5a1b2-486a-11e8-9675-0c33d4f90f46.png)

**Expected outcome**
Notice the empty space between plot and the legend. This was not the case with Matplotlib version: 2.1.0
Below are the plots for the same stuff in both the versions:
Matplotlib 2.1.0:
![matplotlib_2 1 0](https://user-images.githubusercontent.com/5120865/39248713-d20d739c-486a-11e8-9cd1-ce5b2cabac84.PNG)

Matplotb 2.2.2:
![matplotlib_2 2 2](https://user-images.githubusercontent.com/5120865/39248750-e32d8d74-486a-11e8-875f-cbae362f0ab6.PNG)

**Matplotlib version**
  * Operating system: Windows 10
  * Matplotlib version: '2.2.2'
  * Matplotlib backend: Qt5Agg or module://ipykernel.pylab.backend_inline
  * Python version: Python 3.6.5 |Anaconda custom (64-bit)
  * Jupyter Notebook version: 6.3.1

I have the Python installed through Anaconda and I spotted the above change when I updated my packages through Anaconda.
![image](https://user-images.githubusercontent.com/5120865/39249294-58cf95da-486c-11e8-8d4f-ee8c2c22dc8f.png)

**Comment by tacaswell** [comment]
attn @jklymak

**Comment by jklymak** [comment]
Yes - https://matplotlib.org/users/whats_new.html#axes-legends-now-included-in-tight-bbox (#9164)

You have two choices with tight_layout - you can either make people mad because it includes legends, or you can make them mad because it doesn't include legends.  

To use 2.2 you have two choices.  
   1. use fig.legend, in which case you will have to keep or make a list of all the legend items and their labels. 
  2. don't use `tight_layout` and position your axes manually

We discussed this on the call a bit this week - the consensus was to provide a ready handle/label gatherer for passing to `figure.legend`.  i.e. `handles, labels = ax.get_handles_and_labels()`

**Comment by deepak18** [comment]
1) If I don't use ```fig.tight_layout```, the plot is not of the exact size as I mention in ```plt.subplots(figsize=(w,h))``` like:

2) What if I don't use ```fig.transFigure``` in ax.legend() and do something like:
```
plt.legend(bbox_to_anchor=(1, -0.1), loc="upper right",
            columnspacing=0.5, ncol=2)

fig.tight_layout()
```
Then it will be of the size I mention and I can avoid the unnecessary space between plot and legend like:
![image](https://user-images.githubusercontent.com/5120865/39254888-334c05c0-4879-11e8-9a74-644bb06a1289.png)

But, then the thing with putting legend corresponding to axes is that there is no general solution. If xtick-labels are long enough the postion that worked for the above fig won't work anymore like below. Every plot will require a customized positioning. 
![image](https://user-images.githubusercontent.com/5120865/39255164-e0a9b6ea-4879-11e8-885a-8db1e68d4680.png)


Do you know, if I can tackle this in a different way?

**Comment by ImportanceOfBeingErnest** [comment]
Can't you call `tight_layout` first, then create the legend?

**Comment by deepak18** [comment]
@ImportanceOfBeingErnest , that seems to work like a charm. Thank you so much.

Since the legend is also included in ```tight_layout``` it makes sense to fit the plot first and then use the figure coordinates to place the legend. Previously, it just didn't matter.

**Comment by ImportanceOfBeingErnest** [comment]
@jklymak I cannot reproduce the image you show. The legend should be *below* the ticklabels. The strategy of the OP was to position the legend *outside*  the figure (`bbox_to_anchor=(1, 0), loc="upper right",`). Then `tight_layout` will make sure everything but the legend stays within the figure. Hence, no overlap between legend and axes/labels etc. This seems to be a useful strategy in general, because it can be applied without hardcoding any special coordinates and is completely independent on the actual figure content.

**Comment by jklymak** [comment]
Ooops, sorry, my fault, I changed the `upper right` to `lower right` so it would display.  

That way of doing things works if you never want to see the legend on the screen... (i.e. w/ `plt.show()`).  Not a very satisfactory state of affairs in general.  

You could also using 2.2 can do:

```python 
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.transforms as mtransforms

# Fixing random state for reproducibility
np.random.seed(19680801)

N = 50
x = 15*np.random.rand(N)
y = 15*np.random.rand(N)
colors = np.random.rand(N)

fig, ax = plt.subplots(figsize=(8,6))

ax.scatter(x, y, c=colors, alpha=0.5, label='Scatter plot')
ax.plot([0, 15],[0, 15], color='r', linestyle='-', linewidth=2.5, label = 'Straight line')

ax.legend(bbox_to_anchor=(1, 0), loc="lower right",
           bbox_transform=mtransforms.blended_transform_factory(
    ax.transAxes, fig.transFigure), columnspacing=0.5, ncol=2)

fig.tight_layout()

plt.show()
``` 

![boo](https://user-images.githubusercontent.com/1562854/39262986-d928b26c-4875-11e8-9dfe-dc99c38d23fa.png)

**Comment by ImportanceOfBeingErnest** [comment]
Don't forget that a large number of people use matplotlib in jupyter notebooks where the inline backend displays automatically what you would otherwise get via `savefig(.., bbox_inches="tight")`.

**Comment by jklymak** [comment]
cross ref #10682
